backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/VmManagementCommandBase.java

```diff
@@ -116,15 +116,20 @@ protected boolean setAndValidateCpuProfile() {
     }
 
     protected boolean validateCPUHotplug(VmStatic vmStaticData) {
-        if (getVm().isRunningOrPaused()) {
-            // Can't set more CPUs than available on the host where VM is running.
-            // Potential overcommit (interference with other running VMs) will be resolved by scheduler.
-            // In alignment with the CPUPolicyUnit, VM's hyperthreading is not considered.
-            if (getVds() != null && vmStaticData.getNumOfCpus(false) > getVds().getCpuThreads()) {
-                return false;
-            }
+        // Can't set more CPUs than available on the host where VM is running.
+        // Potential overcommit (interference with other running VMs) will be resolved by scheduler.
+        // In alignment with the CPUPolicyUnit, VM's hyperthreading is not considered.
+        if (vmStaticData.getNumOfCpus(false) > getVds().getCpuThreads()) {
```

Comment:
This should be probably changed to check only the number of shared CPUs of the host otherwise we will end up with running VM that has more cores than shared CPUs on the host and I am not sure how that affects calculations in the scheduler.

Reply:
shouldn't we also take into account countThreadsAsCore?

Reply:
I am sorry, I was not clear - I meant if we should consider `countThreadsAsCores` when counting available CPUs on the host. 
```
int availableThreads = getVds().getCpuThreads() - vdsCpuUnitPinningHelper.countUnavailableCpus(
                getVdsManager(getVdsId()).getCpuTopology(), true);
```
Looking into the CpuPolicyUnit it should be something like this:
```
int availableCpus = SlaValidator.getEffectiveCpuCores(host, getCluster().getCountThreadsAsCores()) - vdsCpuUnitPinningHelper.countUnavailableCpus(
                getVdsManager(getVdsId()).getCpuTopology(),  getCluster().getCountThreadsAsCores()));
```

I know it was not done that way before but this part is not connected to scheduler in any way that it might just have been a mistake that was never found because the VM was able to run. After the refactoring due to the exclusive pinning I think we might end up in much worse state. 

I have not tried the scenario, but I can imagine that if the cluster is set with `countThreadsAsCores = false` and we run and hot plug a shared VM with so many CPUs as is host's threads and it will propagate to the host's statistics, that we will not be able to schedule any other shared VM because of the calculation in `CpuPolicyUnit`:
```
hostCpuCount - exclusiveCpus - maxSharedCpuCount < 0
```
E.g. 2 (host cores) - 0 - 4 (host threads and the number of VM's CPU after the hotplugging) < 0 (always)

Reply:
ok, and we will need to change also `vmStaticData.getNumOfCpus(false)` to consider `countThreadsAsCores`

Reply:
I am also thinking how to connect this code with `CpuPolicyUnit` so that they share the same logic. I found that the policies are instantiated in SchedulerManager like this:
```
InternalPolicyUnits.instantiate(CPUPolicyUnit.class, getPendingResourceManager());
```
Maybe we can instantiate the CPUPolicyUnit here and use it to calculate if the VM can run? The calculation would always be identicat and it would include the pending resources too. WDYT? Worth a shot or is it an overkill?

Reply:
> ok, and we will need to change also `vmStaticData.getNumOfCpus(false)` to consider `countThreadsAsCores`

No, not really, `CpuPolicyUnit` counts VM's CPUs as `vmStaticData.getNumOfCpus(false) ` too.

Reply:
> I am also thinking how to connect this code with `CpuPolicyUnit` so that they share the same logic. I found that the policies are instantiated in SchedulerManager like this:
> 
> ```
> InternalPolicyUnits.instantiate(CPUPolicyUnit.class, getPendingResourceManager());
> ```
> 
> Maybe we can instantiate the CPUPolicyUnit here and use it to calculate if the VM can run? The calculation would always be identicat and it would include the pending resources too. WDYT? Worth a shot or is it an overkill?

I think it's overkill to run the whole policy, although it will be the most correct way (the most precise).
The question if it worth it, the big benefit in my opinion is the pending resource. You really need to hit the timing of initiating new VMs and hotplugging. 



> > ok, and we will need to change also `vmStaticData.getNumOfCpus(false)` to consider `countThreadsAsCores`
> 
> No, not really, `CpuPolicyUnit` counts VM's CPUs as `vmStaticData.getNumOfCpus(false) ` too.

OK, so better to align with that. Done.

Reply:
> > I am also thinking how to connect this code with `CpuPolicyUnit` so that they share the same logic. I found that the policies are instantiated in SchedulerManager like this:
> > ```
> > InternalPolicyUnits.instantiate(CPUPolicyUnit.class, getPendingResourceManager());
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe we can instantiate the CPUPolicyUnit here and use it to calculate if the VM can run? The calculation would always be identicat and it would include the pending resources too. WDYT? Worth a shot or is it an overkill?
> 
> I think it's overkill to run the whole policy, although it will be the most correct way (the most precise). The question if it worth it, the big benefit in my opinion is the pending resource. You really need to hit the timing of initiating new VMs and hotplugging. And if we are looking about `CPUPolicyUnit`, why not running `CpuOverloadPolicyUnit` policy as well? which even make it more overkill.

it sounds not easy but I don't think it's an overkill - yes, the likelihood of that to happen is really low but unless I'm missing something, if that happens we may fail to run any vm with dedicated cpu afterwards
but yeah, that's something that can be done separately, no need to fix everything in this PR

Reply:
@ahadas To be clear, It is not related to dedicated, it affects all VMs.
@liranr23 The main difference between `CPUPolicyUnit` and `CpuOverloadPolicyUnit` is that the cpu overload would lead to some overloading issues and would be resolved by ballancer. On the other hand, if we calculate it wrong, we might not be able to run any VM.